File: impls/perl/step1_read_print.pl
#!perl
use strict;
use warnings;
use FindBin;
use lib $FindBin::Bin;
use Term::ReadLine;
use Reader;
use Printer;

sub READ {
    my $str=shift;
    return Reader::read_str($str);
}
sub EVAL {
    my $val=shift;
    return $val;
}
sub PRINT {
    my $val=shift;
    return Printer::pr_str($val);
}

sub rep {
    my $line=shift;
    return PRINT EVAL READ $line;
}

my $line;
my $term=Term::ReadLine->new('mal');
my $prompt='user> ';
my $OUT=$term->OUT || \*STDOUT;
while(defined ($line=$term->readline($prompt))) {
    chomp $line;
    $term->addhistory($line) if $line =~ /\S/;
    eval {
        print $OUT rep($line), "\n";
    } or do {
        print $@, "\n" if $@;
    };
}

File: impls/js/step1_read_print.js
const readline = require('readline');
const read_str = require('./reader.js');
const pr_str = require('./printer.js');

function READ(line) {
    return read_str(line);
}
function EVAL(value) {
    return value;
}
function PRINT(value) {
    return pr_str(value, true);
}
function repl(line) {
    return PRINT(EVAL(READ(line)));
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: 'user> '
});

rl.prompt();

rl.on('line', (line) => {
    try {
        console.log(repl(line));
    } catch (e) {
        console.log(e);
    }
    rl.prompt();
}).on('close', () => {
  process.exit(0);
});
